### PR TITLE
Skipchain leader only

### PR DESCRIPTION
**Pull request by Gilthoniel**
This adds information to the Handler.Process so that network layer info are accessible. It then uses that information to force a skipchain block to be proposed by the leader only.

### CODE REVIEW

File: mino/mod.go [hunk 1]
```diff
@@ -38,6 +39,14 @@ type Sender interface {
 	Send(msg proto.Message, addrs ...Address) <-chan error
 }
 
+// Request is a wrapper around the context of a message received from a player
+// and that needs to be processed by the node. It provides some useful
+// information about the network layer.
+type Request struct {
+	Address Address
```
Comment (nkcr) [inline mino/mod.go]: I would comment what address this corresponds to

File: mino/mod.go [hunk 2]
```diff
@@ -59,7 +68,7 @@ type RPC interface {
 type Handler interface {
 	// Process handles a single request by producing the response according to
 	// the request message.
-	Process(req proto.Message) (resp proto.Message, err error)
+	Process(req Request) (resp proto.Message, err error)
```
Comment (nkcr) [inline mino/mod.go]: following that, shouldn't we also update Receiver.Recv from `Recv(context.Context) (Address, proto.Message, error)` to `Recv(context.Context) (Request, error)` ?

Comment (nkcr) [reply mino/mod.go]: Also, if this is not likely that we add more things to `Request`, I would find it better to try keeping this interface simple and not use a new struct: `Process(req Request, addr Address) (resp proto.Message, err error)`, as you are already doing with Receiver.Recv

Comment (Gilthoniel) [reply mino/mod.go]: As discussed, I think it will be useful later.

File: blockchain/mod.go [hunk 3]
```diff
@@ -29,10 +31,14 @@ type BlockFactory interface {
 	FromVerifiable(src proto.Message) (Block, error)
 }
 
-// Validator is the interface to implement to validate the generic payload
-// stored in the block.
-type Validator interface {
+// PayloadProcessor is the interface to implement to validate the generic
+// payload stored in the block.
+type PayloadProcessor interface {
```
Comment (nkcr) [inline blockchain/mod.go]: makes me think: we should also add the definition of "payload" to our terminologies.

File: blockchain/viewchange/mod.go [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+import (
+	"go.dedis.ch/fabric/blockchain"
+	"go.dedis.ch/fabric/mino"
+)
+
+// ViewChange provides primitives to verify if a participant is allowed to
+// propose a block as the leader. It is also responsible for verifying the
+// integrity of the players of the chain.
+type ViewChange interface {
+	// Wait returns a non-nil error when the node is allowed to make the
```
Comment (nkcr) [inline blockchain/viewchange/mod.go]: - You return an error when you are allowed to make a proposal?
- I would use the same word in the comment and the code: in the comment you speak about "proposal" and in the code its "block". 
- It feels strange to suddenly speak about a "node". I would be a bit more precise by saying maybe "...returns a non-nil error when the node *running the code* is..."

File: mino/option.go [hunk 5]
```diff
@@ -30,6 +30,25 @@ func ApplyFilters(filters []FilterUpdater) *Filter {
 // FilterUpdater is a function to update the filters.
 type FilterUpdater func(*Filter)
 
+// RotateFilter is a filter to rotate the indices. When n is above zero, it will
```
Comment (nkcr) [inline mino/option.go]: so we should update the comment at line 12: "This list is always sorted."

File: blockchain/viewchange/mod.go (hunk 4 above)
Comment (Gilthoniel) [reply blockchain/viewchange/mod.go]: Oops, no it's a nil error when allowed. I changed node by player to be more consistent.

File: mino/option.go (hunk 5 above)
Comment (Gilthoniel) [reply mino/option.go]: There's already the comment ;)

**Pull request merged**